src/storage/mvcc/txn.rs

```diff
@@ -620,6 +626,75 @@ impl<S: Snapshot> MvccTxn<S> {
         })
     }
 
+    /// Check the status of a transaction.
+    ///
+    /// This operation checks whether a transaction has expired it's Lock's TTL, rollback the
+    /// transaction if expired, and update the transaction's min_commit_ts according to the metadata
+    /// in the primary lock.
+    ///
+    /// When transaction T1 meets T2's lock, it may invoke this on T2's primary key. In this
+    /// situation, `self.start_ts` is T2's `start_ts`, `caller_start_ts` is T1's `start_ts`, and
+    /// the `current_ts` is literally the timestamp when this function is invoked. It may not be
+    /// accurate. If `current_ts` is 0, the lock will always be rolled back.
+    ///
+    /// Returns (`lock_ttl`, `commit_ts`, `is_pessimistic_txn`).
+    /// After checking, if the lock is still alive, it retrieves the Lock's TTL; if the transaction
+    /// is committed, get the commit_ts; otherwise, if the transaction is rolled back or there's
+    /// no information about the transaction, results will be both 0.
+    pub fn check_txn_status(
+        &mut self,
+        primary_key: Key,
+        caller_start_ts: u64,
+        current_ts: u64,
+    ) -> Result<(u64, u64, bool)> {
+        match self.reader.load_lock(&primary_key)? {
+            Some(ref mut lock) if lock.ts == self.start_ts => {
+                let is_pessimistic_txn = lock.for_update_ts != 0;
+
+                if current_ts == 0
+                    || extract_physical(lock.ts) + lock.ttl < extract_physical(current_ts)
+                {
+                    // When current_ts == 0 or the lock is expired, clean it up.
+                    if lock.lock_type == LockType::Pessimistic {
+                        self.unlock_key(primary_key);
```

Comment:
Why can we avoid writing Rollback when it's a pessimistic lock? What if there comes a stale acquire command?

Reply:
I didn't write rollback because I found that `pessimistic_rollback` didn't write rollback either. I'm not very understand now. Does `pessimistic_rollback` need to write rollback too?

Reply:
`pessimistic_rollback` is called by the lock owner transaction itself. Other transactions still use `cleanup` to clean pessimistic locks.